libcstr/man3/cstr.3
.\"
.\" Written 2011 by David Herrmann
.\" Dedicated to the Public Domain
.\"
.TH "CSTR" 3 "August 2011" "David Herrmann" "CSTR Library"
.SH NAME
cstr \- Static array and c-string library

.SH SYNOPSIS
.B #include <libcstr.h>

typedef
.B struct cstr
{
.br
	size_t
.B len;
.br
	ssize_t
.B size;
.br
	uint8_t
.B *buf;
.br
}
.B cstr;

#define
.B CSTR_LEN(str)
((str)->len)
.br
#define
.B CSTR_SIZE(str)
(abs((str)->size))
.br
#define
.B CSTR_CHAR(str)
((char*)((str)->buf))
.br
#define
.B CSTR_VOID(str)
((void*)((str)->buf))
.br
#define
.B CSTR_UINT8(str)
((uint8_t*)((str)->buf))

.RB "cstr *" "cstr_alloc" "(size_t len, ssize_t size, void *buf);"
.br
.RB "void " "cstr_clear" "(cstr *str);"
.br
.RB "void " "cstr_free" "(cstr *str);"
.br
.RB "bool " "cstr_fit" "(cstr *str, size_t len);"
.br
bool cstr_cfit(cstr *str, size_t len);
.br
.RB "cstr *" "cstr_new" "(size_t len);"
.br
cstr *cstr_cnew(size_t len);
.br
.RB "cstr *" "cstr_dup" "(const cstr *str);"
.br
cstr *cstr_cdup(const cstr *str);
.br
.RB "bool " "cstr_cat" "(cstr *str, const cstr *cat);"
.br
bool cstr_ccat(cstr *str, const cstr *cat);

.SH DESCRIPTION
The
.B CSTR Library
helps managing static arrays and strings in C. Although not limited to
char-arrays, it is most commonly used for c-strings and hence its name. A cstr
object can be created on the heap or stack. Its internal string buffer can be
dynamically resized by the library or use a statically assigned buffer without
any overhead.

Memory management is handled by the library. Functions that may fail due to
memory allocation failure return false/NULL on error. See each function for
further details.
The library uses standard
.B malloc(), realloc()
and
.B free()
functions for memory management.

A cstr object consists of the array-length, the array-size and the actual array.
The array-length is the length of the array. The array-size is the actually
allocated size of the array. It is always either greater than or equal to the
array-length. If the buffer is static and not allocated by the library, then the
array-size is negative and the buffer will not be freed by the library. Instead,
on reallocation, the buffer is dropped and a new buffer is allocated.
Size 0 is always a static buffer with only a terminating zero character.
Since many strings are modified during runtime, a cstr buffer, when allocated,
has always twice the size as required. This allows to modify the string without
the need of reallocating the buffer. However, often a string is constant and
does not need to be modified. Therefore, constant strings are supported. A
constant string can be used the same way as other strings but its array-size is
always the same as its array-length. So this type does not keep a buffer twice
its size.

Many functions that operate on a strings need to change the string size. This is
always done dynamically if the buffer of the string is not sufficient. If a new
buffer is allocated, it is always twice the size as needed. The buffer never
shrinks, though.
Functions that may reallocate a string always return a boolean indicating
success or failure. The string object is always valid after the function call,
but if the function returned false, then the operation did not modify the
string at all.

Most functions operate on many strings but only one string is the destination of
the operation. If that string is
considered constant and, hence, should not keep a buffer, then the function's
name is prefixed with a
.B c
(for instance cstr_dup() and cstr_cdup(): The latter one will duplicate the
given string with a buffer of the exact string length. The first one will
duplicate the string with a buffer twice the size of the string length)

For backwards compatibility every cstr object's buffer is terminated with a
binary zero-character like classic C-string constants. This character is not
considered part of the string and hence not included in string-length or
string-size. But you can always be sure that cstr->buf[cstr->len] is 0. This
allows to print this string with classic printf()-style functions. However, if
your cstr object contains multiple binary zero-characters inside the string,
then the string will be truncated in the output as printf() is not compatible
with such strings.
If you manage or initialize the buffer yourself and pass it to a new cstr
object, then you must always make the buffer 1 byte bigger than requested. That
is, your "size" argument is 1 smaller than the actual size. Otherwise, the
library won't be able to write the 0 character.
The 0 character is not added to the size argument, as it is just a
backwards-compatibility helper. New applications should not rely on anything
like it when handling cstr objects, except if they pass the cstr object as
classic C-String to other functions.

New strings almost always should be allocated either with
.B cstr_new()
or
.B cstr_dup().
Then you may modify the strings and when you are done, you should destroy them
with
.B cstr_free().

To access the raw data, you should directly access the structure. Use
.B cstr->len
to get the string length and
.B cstr->buf
to get a pointer to the buffer. Some convenience macros are available that also
perform the required casts. See
.B CSTR_LEN, CSTR_VOID, CSTR_CHAR, CSTR_UINT8.
They return
.B cstr->buf
but also convert it to the requested pointer type. Feel free to modify these
values, too. However, the cstr functions are optimized for most string
modifications so it's better to use them.

It is also possible to allocate cstr objects on the stack or as lvalue. See
CSTR_LVALUE for the documentation. This allows to call functions, that
accept cstr-objects like cstr_some_func(CSTR_CS("some string"));

.SH FILES
.I /usr/include/libcstr.h
.RS
The main header file.
.RE
.I /usr/lib/libcstr.so
.RS
The compiled shared library.

.SH BUGS
The namespace "cstr" might clash with other header files.

.SH AUTHOR
David Herrmann <dh.herrmann@googlemail.com>
.SH "SEE ALSO"
.BR cstr_alloc (3),
.BR cstr_free (3),
.BR cstr_fit (3),
.BR cstr_new (3),
.BR cstr_dup (3),
.BR cstr_cat (3),


libcstr/src/cstr.c
/*
 * Static C-strings and arrays
 * Written 2011 by David Herrmann
 * Dedicated to the Public Domain
 */

/*
 * CSTR Objects
 * A cstr object consists of a buffer, the size of the buffer and the length of
 * the string. The length is always smaller or equal to the absolute size of the
 * buffer. The length may be 0.
 * The buffer pointer is always non-NULL and points to a valid buffer. The
 * buffer size is either the absolute size of the buffer minus 1 or its
 * negative. If the size is negative, this means that the buffer is not owned by
 * us and that we should not call free() on it. We simply drop it and allocate a
 * new buffer if we need to resize our buffer.
 * If the size is zero or bigger, it means that our buffer is allocated by us
 * and that we need to free it with free() when no longer used.
 * The actual size of the buffer is abs(obj->size) + 1. We keep a +1 for the
 * terminating zero for backwards compatibility.
 *
 * To speed up string manipulations, the buffer is often twice the size as
 * required. However, many strings do not need to be manipulated and hence do
 * not need a big buffer. Therefore, many functions take as last argument a
 * \constant boolean which should be true if no additional buffer space should
 * be allocated. If it is false, the buffer size is always twice as big as
 * requested. The default value is false.
 */

#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include "libcstr.h"

/*
 * Allocates a new cstr object on the heap.
 * The cstr object will have length \len. If \size is greater than or equal to
 * zero and \buf is NULL, then a new buffer is allocated. If \buf is non-NULL,
 * the buffer is set to \buf and we own it now (that is, we free it if we no
 * longer need it).
 * If \size is smaller than 0, then \buf is taken as constant buffer which we do
 * not own.
 * In all cases, abs(\size) must be greater than or equal to \len.
 * If \buf is non-NULL, then buffer pointed to by \buf must be at least of size
 * abs(\size) + 1. The +1 is important!
 *
 * This returns NULL on memory allocation errors, otherwise it returns the new
 * cstr object.
 */
cstr *cstr_alloc(size_t len, ssize_t size, void *buf)
{
	cstr *str;

	assert(len <= abs(size));

	str = malloc(sizeof(*str));
	if (!str)
		return NULL;

	if (size >= 0 && !buf) {
		str->buf = malloc(size + 1);
		if (!str->buf) {
			free(str);
			return NULL;
		}
	} else {
		assert(buf);
		str->buf = buf;
	}

	str->len = len;
	str->size = size;
	str->buf[len] = 0;

	return str;
}

/*
 * This frees the allocated buffer of \str. It does not free the cstr object.
 * However, the cstr object is invalid after this call and must not be used,
 * anymore. The only valid function that may be called on the cstr object after
 * this is cstr_free().
 * You do not need to call this on dynamically allocated cstr objects as
 * cstr_free() also frees the buffers. However, statically allocated cstr
 * objects which are not constant must be freed with this function.
 */
void cstr_clear(cstr *str)
{
	if (str->size >= 0)
		free(str->buf);
	str->len = 0;
	str->size = 0;
	str->buf = NULL;
}

/*
 * Frees all allocated buffers of \str and the object itself.
 * After this call \str should not be used anymore.
 */
void cstr_free(cstr *str)
{
	if (str) {
		cstr_clear(str);
		free(str);
	}
}

/*
 * Resize cstr buffer
 * This resizes the buffer of \str to make room for at least \len bytes. The
 * string length is also reset to \len and the zero terminating character is set
 * to the end of the new buffer.
 * If the current buffer is big enough, no new buffer is allocated. Only if we
 * need more space, the current buffer is freed or dropped and a new buffer is
 * allocated.
 * This returns false if the new buffer cannot be allocated. Otherwise it
 * returns true.
 * If a new buffer is allocated and \constant is true, the new buffer will have
 * the exact same size as required. If \constant is false, the buffer will be
 * twice as big as required.
 * \str is not touched at all if the memory allocation fails and false is
 * returned. So on failure the old state is preserved.
 */
bool cstr__fit(cstr *str, size_t len, bool constant)
{
	void *snew;
	size_t size;

	assert(str);

	if (abs(str->size) < (ssize_t)len) {
		if (constant)
			size = len;
		else
			size = len * 2;

		if (str->size < 0)
			snew = malloc(size + 1);
		else
			snew = realloc(str->buf, size + 1);
		if (!snew)
			return false;
		str->buf = snew;
		str->size = size;
	}

	str->len = len;
	str->buf[len] = 0;

	return true;
}

/*
 * Duplicate string
 * This creates a duplicate of \old and returns it. The new object is in no way
 * dependent of \old and can be used as usual.
 * Returns NULL on memory allocation errors.
 * The buffer of the new object may be of different size as \old. \constant
 * specifies how the new buffer is allocated.
 */
cstr *cstr__dup(const cstr *old, bool constant)
{
	cstr *dest;

	assert(old);

	if (constant)
		dest = cstr_cnew(CSTR_LEN(old));
	else
		dest = cstr_new(CSTR_LEN(old));
	if (!dest)
		return NULL;

	memcpy(CSTR_VOID(dest), CSTR_VOID(old), CSTR_LEN(old));
	return dest;
}

/*
 * Concatenate strings
 * This puts \src at the end of \dest. \constant controls how the buffer of
 * \dest is reallocated if required.
 */
bool cstr__cat(cstr *dest, const cstr *src, bool constant)
{
	size_t dlen = CSTR_LEN(dest);

	if (!cstr__fit(dest, dlen + CSTR_LEN(src), constant))
		return false;

	memcpy(CSTR_UINT8(dest) + dlen, CSTR_VOID(src), CSTR_LEN(src));
	return true;
}

/*
 * Compare strings
 * Returns true if \str1 and \str2 are equal.
 */
bool cstr_cmp(const cstr *str1, const cstr *str2)
{
	if (CSTR_LEN(str1) != CSTR_LEN(str2))
		return false;

	return !memcmp(CSTR_VOID(str1), CSTR_VOID(str2), CSTR_LEN(str1));
}

/*
 * Compare substrings
 * Compares the first \n characters of both strings for equality.
 * Returns true if they are equal.
 * If \n is bigger than the length of any of both strings, then this is
 * identical to cstr_cmp().
 */
bool cstr_ncmp(const cstr *str1, const cstr *str2, size_t n)
{
	if (CSTR_LEN(str1) < n || CSTR_LEN(str2) < n)
		return cstr_cmp(str1, str2);

	return !memcmp(CSTR_VOID(str1), CSTR_VOID(str2), n);
}

/*
 * Copy string
 * This copies \src into \dest. \constant controls how the new buffer is
 * allocated.
 * This function is similar to cstr_dup() but takes an existing object instead
 * of creating a new object.
 * Returns true on success and false on memory allocation failure.
 */
bool cstr__cpy(cstr *dest, const cstr *src, bool constant)
{
	if (!cstr__fit(dest, CSTR_LEN(src), constant))
		return false;

	memcpy(CSTR_VOID(dest), CSTR_VOID(src), CSTR_LEN(src));
	return true;
}
